### ISSUE TITLE
Loss calculation incorrect?

### ISSUE THREAD

**Issue by kushalj001** [facebookresearch/llama-recipes#68]
Hi,
[In the loss calculation](https://github.com/facebookresearch/llama-recipes/blob/905f633dab92688b0a989f8d5cd11d86f882f534/utils/train_utils.py#L93), we are dividing the total loss by the length of the entire dataset. Since HF by default returns a cross-entropy loss with `reduction=mean`, should we not divide by the length of the train data loader instead of the number of examples (dataset length)? I think we're currently dividing the loss twice. Let me know if this is wrong. 
```
loss = model(**batch).loss
loss = loss / gradient_accumulation_steps
total_loss += loss.detach().float()
first_key = next(iter(batch))
data_set_len += len(batch[first_key])

...

train_epoch_loss = total_loss / data_set_len
```

**Comment by HamidShojanazeri** [comment]
@kushalj001 thanks for opening this issue, the fix will be included in this [PR](https://github.com/facebookresearch/llama-recipes/pull/51).

**Comment by kushalj001** [comment]
There's also two scheduler updates in training: [here](https://github.com/facebookresearch/llama-recipes/blob/a2403c7c1af5e852ac3a3262f5c929b7197bd2d5/utils/train_utils.py#L135) and [here](https://github.com/facebookresearch/llama-recipes/blob/2243b962fa98c9c344f8f1a62be771bb1b71c56f/utils/train_utils.py#L123). I am not sure if this is intentional or a bug.

**Comment by thtang** [comment]
What's the difference or concern of using customize train_utils instead of huggingface trainer?

**Comment by HamidShojanazeri** [comment]
> There are also two LR scheduler updates in training: [here](https://github.com/facebookresearch/llama-recipes/blob/2243b962fa98c9c344f8f1a62be771bb1b71c56f/utils/train_utils.py#L162) and [here](https://github.com/facebookresearch/llama-recipes/blob/2243b962fa98c9c344f8f1a62be771bb1b71c56f/utils/train_utils.py#L123). I am not sure if this is intentional or a bug.

Thanks @kushalj001 its been fixed in the same PR.

**Comment by HamidShojanazeri** [comment]
> What's the difference or concern of using customize train_utils instead of huggingface trainer?

This gives us more flexibility to use Pytorch APIs and pull the latest on FSDP without having to upstream all changes to HF trainer before landing here.

**Comment by thtang** [comment]
> > What's the difference or concern of using customize train_utils instead of huggingface trainer?
> 
> This gives us more flexibility to use Pytorch APIs and pull the latest on FSDP without having to upstream all changes to HF trainer before landing here.

Got it thanks!
